File: android/app/src/main/java/com/ciro/app/data/model/Resource.kt
package com.ciro.app.data.model

/**
 * Kotlin data class mirroring the Firestore `resources` collection schema.
 * TEAM CONTRACT — matches the Python Pydantic Resource model exactly.
 */
data class Resource(
    val resource_id: String = "",
    val type: String = "",
    val state: String = "AVAILABLE",
    val location: ResourceLocation = ResourceLocation(),
    val name: String = "",
    val assigned_incident_id: String? = null,
    val eta_minutes: Double? = null,
    val capacity: Int = 1,
    val unit_name: String = "",
    val contact: String = "",
    val last_updated: String = "",
) {
    /** Emoji icon for map markers based on resource type. */
    val typeIcon: String
        get() = when (type) {
            "ambulance" -> "🚑"
            "police_unit" -> "🚔"
            "rescue_team" -> "⛑️"
            "water_tanker" -> "🚒"
            "field_team" -> "👷"
            "shelter" -> "🏠"
            "generator" -> "⚡"
            "drone" -> "🛸"
            else -> "📍"
        }

    val isAvailable: Boolean get() = state == "AVAILABLE"
    val isDispatched: Boolean get() = state == "DISPATCHED"
    val isShadowCommitted: Boolean get() = state == "SHADOW_COMMITTED"
}

data class ResourceLocation(
    val lat: Double = 0.0,
    val lng: Double = 0.0,
    val name: String = "",
)

File: ciro/schemas/resource.py
# schemas/resource.py
from pydantic import BaseModel, Field
from typing import Optional, Any
from datetime import datetime
import uuid


class ResourceLocation(BaseModel):
    lat: float
    lng: float


class Resource(BaseModel):
    resource_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    type: str
    state: str = "AVAILABLE"
    location: ResourceLocation
    name: str = ""
    assigned_incident_id: Optional[str] = None
    eta_minutes: Optional[float] = None
    capacity: int = 1
    unit_name: str
    contact: str = ""
    last_updated: str = Field(default_factory=lambda: datetime.utcnow().isoformat())

    @property
    def current_location(self):
        from types import SimpleNamespace
        return SimpleNamespace(lat=self.location.lat, lng=self.location.lng, name=self.name)

    def to_firestore_dict(self) -> dict[str, Any]:
        data = self.model_dump()
        data["last_updated"] = datetime.utcnow().isoformat()
        return data

    @classmethod
    def from_firestore_dict(cls, data: dict[str, Any]) -> "Resource":
        return cls(**data)
